Create clear entry points for color parsing to allow uniformly passing options/tracking nesting

weinig opened the pull request:
#### 3d1a81f05a45ba46e924c411cfe06f510109054b
<pre>
Create clear entry points for color parsing to allow uniformly passing options/tracking nesting
<a href="https://bugs.webkit.org/show_bug.cgi?id=273019">https://bugs.webkit.org/show_bug.cgi?id=273019</a>

Reviewed by Darin Adler.

In preparation for additional changes to color parsing, this change
creates a consistent interface for the color parsing/consuming functions,
extracting out the color specific options into a new CSSColorParsingOptions
struct and utilizing the clear entry points to track nesting levels.
(For an example of nesting, parsing `color-mix()` is itself a color,
but as part of its parse, it recursively requires an additional &apos;origin&apos;
color to be parsed).

The nesting level will be useful in the near future to allow changes
in parsing when not at the top level.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
    In clarify the recursion, I fixed a bug where we were unnecessarily
    restrictive and were preventing nested `color-mix()` from working.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::ColorParserState::ColorParserState):
    Mutable state used only in the cpp file to track options and nesting level.

(WebCore::CSSPropertyParserHelpers::ColorParserStateNester::ColorParserStateNester):
(WebCore::CSSPropertyParserHelpers::ColorParserStateNester::~ColorParserStateNester):
    Helper to increment/decrement nesting level using RAII.

(WebCore::CSSPropertyParserHelpers::consumeOriginColorRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorContrastFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::consumeColorMixComponentRaw):
(WebCore::CSSPropertyParserHelpers::consumeColorMixComponent):
(WebCore::CSSPropertyParserHelpers::parseColorMixFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorMixFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseLightDarkFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseHexColor):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunction):
(WebCore::CSSPropertyParserHelpers::consumeColor):
(WebCore::CSSPropertyParserHelpers::consumeColorRaw):
(WebCore::CSSPropertyParserHelpers::parseColorRawWorkerSafe):
(WebCore::CSSPropertyParserHelpers::parseColorRaw):
    Pass ColorParserState down the stack.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Number.cpp:
(WebCore::CSSPropertyParserHelpers::validatedNumberRaw):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Number.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.cpp:
(WebCore::CSSPropertyParserHelpers::validatedPercentRaw):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.h:
    Add validate function forward declarations into the headers. This
    was only compiling due to the unified build happening to group
    things just right.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserWorkerSafe::parseColor): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h:
    Move to CSSPropertyParserConsumer+Color.h.

* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/CustomPaintCanvas.h:
(WebCore::CustomPaintCanvas::cssParserContext const):
* Source/WebCore/html/HTMLCanvasElement.h:
(WebCore::HTMLCanvasElement::cssParserContext const):
(WebCore::HTMLCanvasElement::cssParserContext): Deleted.
* Source/WebCore/html/OffscreenCanvas.h:
(WebCore::OffscreenCanvas::cssParserContext const):
    Since a CSSParserContext is needed for all canvases, move the
    existing HTMLCanvasElement function to be a virtual function
    on CanvasBase and implement for OffscreenCanvas and CustomPaintCanvas.

* Source/WebCore/html/canvas/CanvasStyle.cpp:
(WebCore::parseColor):
(WebCore::currentColor):
    Call the new CSSPropertyParserConsumer+Color.h parse functions. In
    a subsequent change, these will need to pass additional options for
    overriding hsl()/hsla() parsing behavior.

Canonical link: <a href="https://commits.webkit.org/277783@main">https://commits.webkit.org/277783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f1ac463650ac4583e0737e2137b9a37d7857184

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48557 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27768 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51244 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44622 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50862 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33704 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25295 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39704 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49139 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25437 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41891 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20801 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22914 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43064 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6613 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44852 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43553 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53151 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23603 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19905 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46996 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24869 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42097 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45919 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10707 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25673 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24591 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
```diff
@@ -768,7 +777,7 @@ PASS Property color value 'lab(from lab(none none none / none) l a b / alpha)'
 PASS Property color value 'lab(from lab(25 20 50 / none) l a b / alpha)'
 PASS Property color value 'color-mix(in lab, lab(from lab(25 20 50) none a b), lab(25 20 50))'
-FAIL Property color value 'lab(from color-mix(in lab, lab(25 20 50), lab(25 20 50)) l a b / alpha)' assert_true: 'lab(from color-mix(in lab, lab(25 20 50), lab(25 20 50)) l a b / alpha)' is a supported value for color. expected true got false
+PASS Property color value 'lab(from color-mix(in lab, lab(25 20 50), lab(25 20 50)) l a b / alpha)'
```
From EWS, looks like at least two other tests are now passing and need expected results updated:
```
imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.fillStyle.toStringFunctionCallback.html
imported/w3c/web-platform-tests/html/canvas/offscreen/fill-and-stroke-styles/2d.fillStyle.toStringFunctionCallback.worker.html
```

weinig replied:
Nice, updated.

darinadler commented on Source/WebCore/html/OffscreenCanvas.h:
```diff
@@ -29,6 +29,7 @@
 
 #include "ActiveDOMObject.h"
 #include "AffineTransform.h"
+#include "CSSParserContext.h"
```
Would be nice to avoid this include if it’s not needed.

weinig replied:
Done.

darinadler commented on Source/WebCore/html/OffscreenCanvas.h:
```diff
@@ -191,8 +194,18 @@ class OffscreenCanvas final : public ActiveDOMObject, public RefCounted<Offscree
+    mutable std::unique_ptr<CSSParserContext> m_cssParserContext;
 };
 
+inline const CSSParserContext& OffscreenCanvas::cssParserContext() const
```
Given our use of LTO on most platforms, not sure this needs to be in the header.

weinig replied:
Done.

Commit:
Create clear entry points for color parsing to allow uniformly passing options/tracking nesting
https://bugs.webkit.org/show_bug.cgi?id=273019

Reviewed by Darin Adler.

In preparation for additional changes to color parsing, this change
creates a consistent interface for the color parsing/consuming functions,
extracting out the color specific options into a new CSSColorParsingOptions
struct and utilizing the clear entry points to track nesting levels.
(For an example of nesting, parsing `color-mix()` is itself a color,
but as part of its parse, it recursively requires an additional 'origin'
color to be parsed).

The nesting level will be useful in the near future to allow changes
in parsing when not at the top level.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
    In clarify the recursion, I fixed a bug where we were unnecessarily
    restrictive and were preventing nested `color-mix()` from working.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
(WebCore::CSSPropertyParserHelpers::ColorParserState::ColorParserState):
    Mutable state used only in the cpp file to track options and nesting level.

(WebCore::CSSPropertyParserHelpers::ColorParserStateNester::ColorParserStateNester):
(WebCore::CSSPropertyParserHelpers::ColorParserStateNester::~ColorParserStateNester):
    Helper to increment/decrement nesting level using RAII.

(WebCore::CSSPropertyParserHelpers::consumeOriginColorRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRGBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHSLParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseHWBParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseLabParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseNonRelativeLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseLCHParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseRelativeColorFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorContrastFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::consumeColorMixComponentRaw):
(WebCore::CSSPropertyParserHelpers::consumeColorMixComponent):
(WebCore::CSSPropertyParserHelpers::parseColorMixFunctionParametersRaw):
(WebCore::CSSPropertyParserHelpers::parseColorMixFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseLightDarkFunctionParameters):
(WebCore::CSSPropertyParserHelpers::parseHexColor):
(WebCore::CSSPropertyParserHelpers::parseColorFunctionRaw):
(WebCore::CSSPropertyParserHelpers::parseColorFunction):
(WebCore::CSSPropertyParserHelpers::consumeColor):
(WebCore::CSSPropertyParserHelpers::consumeColorRaw):
(WebCore::CSSPropertyParserHelpers::parseColorRawWorkerSafe):
(WebCore::CSSPropertyParserHelpers::parseColorRaw):
    Pass ColorParserState down the stack.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Number.cpp:
(WebCore::CSSPropertyParserHelpers::validatedNumberRaw):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Number.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.cpp:
(WebCore::CSSPropertyParserHelpers::validatedPercentRaw):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.h:
    Add validate function forward declarations into the headers. This
    was only compiling due to the unified build happening to group
    things just right.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserWorkerSafe::parseColor): Deleted.
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h:
    Move to CSSPropertyParserConsumer+Color.h.

* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/CustomPaintCanvas.h:
(WebCore::CustomPaintCanvas::cssParserContext const):
* Source/WebCore/html/HTMLCanvasElement.h:
(WebCore::HTMLCanvasElement::cssParserContext const):
(WebCore::HTMLCanvasElement::cssParserContext): Deleted.
* Source/WebCore/html/OffscreenCanvas.h:
(WebCore::OffscreenCanvas::cssParserContext const):
    Since a CSSParserContext is needed for all canvases, move the
    existing HTMLCanvasElement function to be a virtual function
    on CanvasBase and implement for OffscreenCanvas and CustomPaintCanvas.

* Source/WebCore/html/canvas/CanvasStyle.cpp:
(WebCore::parseColor):
(WebCore::currentColor):
    Call the new CSSPropertyParserConsumer+Color.h parse functions. In
    a subsequent change, these will need to pass additional options for
    overriding hsl()/hsla() parsing behavior.

Canonical link: https://commits.webkit.org/277783@main
